mysql/infoschema.go

```diff
@@ -0,0 +1,364 @@
+	primaryKeys, constraints, err := getConstraints(conv, db, table)
+	if err != nil {
+		return fmt.Errorf("couldn't get constraints for table %s.%s: %s", table.schema, table.name, err)
+	}
+	colDefs, colNames := processColumns(conv, cols, constraints)
+	name := buildTableName(table.schema, table.name)
+	var schemaPKeys []schema.Key
+	for _, k := range primaryKeys {
+		schemaPKeys = append(schemaPKeys, schema.Key{Column: k})
+	}
+	conv.SrcSchema[name] = schema.Table{
+		Name:        name,
+		ColNames:    colNames,
+		ColDefs:     colDefs,
+		PrimaryKeys: schemaPKeys}
+	return nil
+}
+
+func getColumns(table schemaAndName, db *sql.DB) (*sql.Rows, error) {
+	q := `SELECT c.column_name, c.data_type, c.column_type, c.is_nullable, c.column_default, c.character_maximum_length, c.numeric_precision, c.numeric_scale
+              FROM information_schema.COLUMNS c
+              where table_schema = ? and table_name = ? ORDER BY c.ordinal_position;`
+	return db.Query(q, table.schema, table.name)
+}
+
+func processColumns(conv *internal.Conv, cols *sql.Rows, constraints map[string][]string) (map[string]schema.Column, []string) {
+	colDefs := make(map[string]schema.Column)
+	var colNames []string
+	var colName, dataType, isNullable, columnType string
+	var colDefault sql.NullString
+	var charMaxLen, numericPrecision, numericScale sql.NullInt64
+	for cols.Next() {
+		err := cols.Scan(&colName, &dataType, &columnType, &isNullable, &colDefault, &charMaxLen, &numericPrecision, &numericScale)
+		if err != nil {
+			fmt.Printf("Can't scan: %v\n", err)
+			continue
+		}
+		unique := false
+		ignored := schema.Ignored{}
+		for _, c := range constraints[colName] {
+			// c can be UNIQUE, PRIMARY KEY, FOREIGN KEY or CHECK
+			// We've already filtered out PRIMARY KEY.
+			switch c {
+			case "UNIQUE":
+				unique = true
+			case "FOREIGN KEY":
+				ignored.ForeignKey = true
+			case "CHECK":
+				ignored.Check = true
+			}
+		}
+		ignored.Default = colDefault.Valid
+		c := schema.Column{
+			Name:    colName,
+			Type:    toType(dataType, columnType, charMaxLen, numericPrecision, numericScale),
+			NotNull: toNotNull(conv, isNullable),
+			Unique:  unique,
+			Ignored: ignored,
+		}
+		colDefs[colName] = c
+		colNames = append(colNames, colName)
+	}
+	return colDefs, colNames
+}
+
+// getConstraints returns a list of primary keys and by-column map of
+// other constraints.  Note: we need to preserve ordinal order of
+// columns in primary key constraints.
+func getConstraints(conv *internal.Conv, db *sql.DB, table schemaAndName) ([]string, map[string][]string, error) {
+	q := `SELECT k.COLUMN_NAME, t.CONSTRAINT_TYPE
+              FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS t
+                INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS k
+                  ON t.CONSTRAINT_NAME = k.CONSTRAINT_NAME AND t.CONSTRAINT_SCHEMA = k.CONSTRAINT_SCHEMA AND t.TABLE_NAME=k.TABLE_NAME
+              WHERE k.TABLE_SCHEMA = ? AND k.TABLE_NAME = ? ORDER BY k.ordinal_position;`
```

Comment:
Why use '?' instead of $1, $2 for the parameters?

Reply:
MySQL does not support bind parameters `$1,$2..` like postgres. It throws an error when such bind parameters are used:
```
panic: couldn't get schema for table testdb.Persons: Error 1054: Unknown column '$1' in 'where clause'
```
Thus, used `?` instead of `$1,$2..`.

Reply:
Ah right. Got it. That's super annoying.
I thought mysql did support $1, $2 etc.
Maybe its an issue with the golang database/sql driver?

Reply:
Yeah, maybe it's not supported by golang driver.